crypto/rsa/rsa_gen.c

```diff
@@ -224,12 +226,23 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, int primes, BIGNUM *e_value,
         nl = BN_num_bits(r1);
         /*
          * if |r1|, product of factors so far, is not long enough,
-         * re-generate the last prime
+         * or the first 4 bits are less than 0x9, re-generate the last prime.
          *
          * NOTE: This actually can't happen in two-prime case, because of
-         * the way factors are generated.
+         * the way factors are generated. But for multi-prime case, the
+         * modulus could start less than 0x9, which could be utilized to
+         * distinguish a multi-prime private key by using the modulus
+         * in a certificate.
          */
-        if (nl != bitse) {
+        nlb = (nl + 7) / 8;
+        bin = OPENSSL_malloc(nlb);
+        if (bin == NULL)
+            goto err;
+        (void)BN_bn2bin(r1, bin);
+        bint = bin[0] >> 4;
```

Comment:
This assumes the the requested bits to be a multiple of 8.
You should BN_rshift r1 down by bitse-4 so that BN_get_word
is in the range 9..15, maybe the replace the whole nl != bitse 
with that condition on the shifted value

Reply:
Ohhh